### PR TITLE
qq: add update script

### PR DESCRIPTION
**Pull request by fee1-dead**
###### Description of changes

Added an update script for the `qq` package.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/instant-messengers/qq/update.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl nix-prefetch-git jq
+
+set -euo pipefail
+
```
Comment (linsui) [inline pkgs/applications/networking/instant-messengers/qq/update.sh:5]:
```suggestion

cd $(readlink -e $(dirname "${BASH_SOURCE[0]}"))

```

When running `nix-shell maintainers/scripts/update.nix --argstr package qq` the work dir is the root dir.

Comment (fee1-dead) [reply pkgs/applications/networking/instant-messengers/qq/update.sh:5]: done, thanks!

**Review by linsui — APPROVED** [review]

File: pkgs/applications/networking/instant-messengers/qq/update.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -i bash -p curl nix-prefetch-git jq
```
Comment (pbsds) [inline pkgs/applications/networking/instant-messengers/qq/update.sh:2]:
```suggestion
#!nix-shell -i bash --pure -p git nix curl nix-prefetch-git jq
```

File: pkgs/applications/networking/instant-messengers/qq/update.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+amd64_hash=$(nix hash to-sri --type sha256 "$amd64_hash")
+arm64_hash=$(nix hash to-sri --type sha256 "$arm64_hash")
+
+sed -i -E -e "/_amd64.deb\";$/{N
+s#hash = \".*\"#hash = \"$amd64_hash\"#
+}" default.nix
+
+sed -i -E -e "/_arm64.deb\";$/{N
+s#hash = \".*\"#hash = \"$arm64_hash\"#
+}" default.nix
```
Comment (pbsds) [inline pkgs/applications/networking/instant-messengers/qq/update.sh:46]: While well written, i'd say it be much more robust to dump these values into a json or separate nix file which is then loaded/imported into the derivation

File: pkgs/applications/networking/instant-messengers/qq/update.sh (hunk 2 above)
Comment (pbsds) [reply pkgs/applications/networking/instant-messengers/qq/update.sh:2]: I'm not sure if `--pure` will shadow `UPDATE_NIX_ATTR_PATH`, but git and nix are at least needed.

**Review by pbsds — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/qq/update.sh (hunk 3 above)
Comment (pbsds) [reply pkgs/applications/networking/instant-messengers/qq/update.sh:46]: or, consider `common-updater-scripts`

**Review by pbsds — APPROVED** [review]
Tested from two different working dirs, LGTM

File: pkgs/applications/networking/instant-messengers/qq/update.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+payload=$(curl https://im.qq.com/rainbow/linuxQQDownload | grep -oP "var params= \K\{.*\}(?=;)")
+amd64_url=$(jq -r .x64DownloadUrl.deb <<< "$payload")
+arm64_url=$(jq -r .armDownloadUrl.deb <<< "$payload")
+
+urlhash=$(grep -oP "(?<=QQNT/)[a-e0-9]+(?=/linuxqq)" <<< "$amd64_url")
+version=$(grep -oP "(?<=/linuxqq_).*(?=_amd64.deb)" <<< "$amd64_url")
+
+sed -i -E -e "s#version = \".*\"#version = \"$version\"#" default.nix
+sed -i -E -e "s#_hash = \".*\"#_hash = \"$urlhash\"#" default.nix
```
Comment (linyinfeng) [inline pkgs/applications/networking/instant-messengers/qq/update.sh:16]: These two `sed -i` are useless with `source.nix`, `default.nix` should remain unchanged after the update.

**Review by linyinfeng — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/instant-messengers/qq/update.sh (hunk 4 above)
Comment (fee1-dead) [reply pkgs/applications/networking/instant-messengers/qq/update.sh]: Removed, thanks!

**Review by linyinfeng — APPROVED** [review]

**Pull request merged**